x-pack/plugins/apm/server/lib/services/get_services/get_services_items.ts

```diff
@@ -3,88 +3,80 @@
-            },
-            events: {
-              terms: { field: PROCESSOR_EVENT },
-            },
-            environments: {
-              terms: { field: SERVICE_ENVIRONMENT },
-            },
-          },
-        },
-      },
-    },
-  });
-
-  const resp = await client.search(params);
-  const aggs = resp.aggregations;
-
-  const serviceBuckets = aggs?.services.buckets || [];
-
-  const items = serviceBuckets.map((bucket) => {
-    const eventTypes = bucket.events.buckets;
+export type ServicesItemsSetup = Setup & SetupTimeRange & SetupUIFilters;
+export type ServicesItemsProjection = ReturnType<typeof getServicesProjection>;
 
-    const transactions = eventTypes.find((e) => e.key === 'transaction');
-    const totalTransactions = transactions?.doc_count || 0;
+export async function getServicesItems(setup: ServicesItemsSetup) {
+  const params = {
+    projection: getServicesProjection({ setup, noEvents: true }),
+    setup,
+  };
 
-    const errors = eventTypes.find((e) => e.key === 'error');
-    const totalErrors = errors?.doc_count || 0;
+  const [
+    transactionDurationAvg,
+    agentName,
+    transactionRate,
+    errorRate,
+    environments,
+  ] = await Promise.all([
+    getTransactionDurationAvg(params),
+    getAgentName(params),
+    getTransactionRate(params),
+    getErrorRate(params),
+    getEnvironments(params),
+  ]);
 
-    const deltaAsMinutes = (end - start) / 1000 / 60;
-    const transactionsPerMinute = totalTransactions / deltaAsMinutes;
-    const errorsPerMinute = totalErrors / deltaAsMinutes;
+  const allMetrics = [
+    transactionDurationAvg,
+    agentName,
+    transactionRate,
+    errorRate,
+    environments,
+  ];
 
-    const environmentsBuckets = bucket.environments.buckets;
-    const environments = environmentsBuckets.map(
-      (environmentBucket) => environmentBucket.key as string
-    );
+  const serviceNames = uniq(
+    arrayUnionToCallable(
+      allMetrics.flatMap((metric) =>
+        arrayUnionToCallable(metric).map((service) => service.name)
+      )
+    )
+  );
 
+  const items = serviceNames.map((serviceName) => {
     return {
-      serviceName: bucket.key as string,
-      agentName: bucket.agents.buckets[0]?.key as string | undefined,
-      transactionsPerMinute,
-      errorsPerMinute,
-      avgResponseTime: bucket.avg.value,
-      environments,
+      serviceName,
+      agentName:
+        agentName.find((service) => service.name === serviceName)?.value ??
+        null,
+      transactionsPerMinute:
+        transactionRate.find((service) => service.name === serviceName)
+          ?.value ?? 0,
+      errorsPerMinute:
+        errorRate.find((service) => service.name === serviceName)?.value ?? 0,
+      avgResponseTime:
+        transactionDurationAvg.find((service) => service.name === serviceName)
+          ?.value ?? null,
+      environments:
+        environments.find((service) => service.name === serviceName)?.value ??
+        [],
```

Comment:
This suddenly got a lot more complex. Any reason why `transactionRate` and the others are now arrays instead of scalar values?
Shouldn't `transactionRate` be a numeric value in the first place, and not an array?

Reply:
Now using `joinByKey` to make this code easier to understand.